File: triangles.ts
#!/usr/bin/env bun

import {
  Bool,
  Nat,
  Real,
  Vec,
  add,
  and,
  compile,
  div,
  fn,
  geq,
  gt,
  iadd,
  ieq,
  ileq,
  ilt,
  leq,
  lt,
  mul,
  neg,
  not,
  or,
  select,
  sqrt,
  sub,
  vec,
  vjp,
} from "rose";
import seedrandom from "seedrandom";
import * as optimizer from "./optimizer.js";

const sqr = (x: Real): Real => mul(x, x);

const max = (a: Real, b: Real) => select(gt(a, b), Real, a, b);

const min = (a: Real, b: Real): Real => select(lt(a, b), Real, a, b);

const clamp = (x: Real, l: Real, h: Real): Real =>
  select(lt(x, l), Real, l, select(lt(h, x), Real, h, x));

const all = (xs: Bool[]): Bool => xs.reduce((a, b) => and(a, b));

const vnot = (xs: Bool[]): Bool[] => xs.map((x) => not(x));

const vadd = (a: Vec<Real> | Real[], b: Vec<Real> | Real[]): Real[] => {
  return [add(a[0], b[0]), add(a[1], b[1])];
};

const vsub = (a: Vec<Real> | Real[], b: Vec<Real> | Real[]): Real[] => {
  return [sub(a[0], b[0]), sub(a[1], b[1])];
};

const vmul = (v: Real[], c: Real): Real[] => v.map((x) => mul(x, c));

const dot = (a: Real[], b: Real[]): Real =>
  a.map((x, i) => mul(x, b[i])).reduce((a, b) => add(a, b));

const norm = (v: Real[]): Real => sqrt(dot(v, v));

const cross = (a: Real[], b: Real[]): Real =>
  sub(mul(a[0], b[1]), mul(a[1], b[0]));

const reorder = fn([Vec(3, Vec(2, Real))], Vec(5, Vec(2, Real)), (p) => {
  let i: Nat = 0;
  for (let j = 1; j < 3; ++j) {
    const [xi, yi] = p[i] as any;
    const [xj, yj] = p[j];
    i = select(or(lt(yj, yi), and(leq(yj, yi), lt(xj, xi))), 3, j, i);
  }
  const [a, b, c] = p;
  return select(
    ieq(3, i, 0),
    Vec(5, Vec(2, Real)),
    [a, b, c, a, b],
    select(
      ieq(3, i, 1),
      Vec(5, Vec(2, Real)),
      [b, c, a, b, c],
      [c, a, b, c, a],
    ),
  );
});

// note: only works on triangles
const minkowskiSum = (
  left: Vec<Vec<Real>>,
  right: Vec<Vec<Real>>,
): Real[][] => {
  const p = reorder(left);
  const q = reorder(right);
  const r: Real[][] = [];
  let i: Nat = 0;
  let j: Nat = 0;
  for (let k = 0; k < 6; ++k) {
    r.push(vadd(p[i], q[j]));
    const i1 = iadd(5, i, 1);
    const j1 = iadd(5, j, 1);
    const z = cross(vsub(p[i1], p[i]), vsub(q[j1], q[j]));
    ({ i, j } = select(
      and(ilt(5, i, 3), geq(z, 0)),
      { i: 5, j: 5 },
      { i: i1, j },
      { i, j: j1 },
    ));
  }
  return r;
};

// https://iquilezles.org/articles/distfunctions2d/
const sdPolygon = (v: Real[][], p: Real[]): Real => {
  const N = v.length;
  let d = dot(vsub(p, v[0]), vsub(p, v[0]));
  let s: Real = 1.0;
  for (let i = 0, j = N - 1; i < N; j = i, i++) {
    const e = vsub(v[j], v[i]);
    const w = vsub(p, v[i]);
    const b = vsub(w, vmul(e, clamp(div(dot(w, e), dot(e, e)), 0.0, 1.0)));
    d = min(d, dot(b, b));
    const c = [
      geq(p[1], v[i][1]),
      lt(p[1], v[j][1]),
      gt(mul(e[0], w[1]), mul(e[1], w[0])),
    ];
    s = select(or(all(c), all(vnot(c))), Real, neg(s), s);
  }
  return mul(s, sqrt(d));
};

type Vec2 = number[];

interface Triangle {
  a: Vec2;
  b: Vec2;
  c: Vec2;
}

const area = (ab: Real, bc: Real, ca: Real): Real => {
  const one = add(add(ab, bc), ca);
  const two = add(add(neg(ab), bc), ca);
  const three = add(sub(ab, bc), ca);
  const four = sub(add(ab, bc), ca);

  return mul(0.25, sqrt(mul(mul(mul(one, two), three), four)));
};

const clockwise = (a: Real[], b: Real[], c: Real[]): Bool =>
  lt(
    mul(sub(b[0], a[0]), sub(c[1], a[1])),
    mul(sub(c[0], a[0]), sub(b[1], a[1])),
  );

const numTriangles = 10;

const fanout = fn([Real], Vec(numTriangles, Real), (x) =>
  vec(numTriangles, Real, () => x),
);

const sum = fn([Vec(numTriangles, Real)], Real, (xs) =>
  vjp(fanout)(0).grad(xs),
);

const size = 100;

const lagrangian = fn(
  [
    {
      ax: Vec(numTriangles, Real),
      ay: Vec(numTriangles, Real),
      bx: Vec(numTriangles, Real),
      by: Vec(numTriangles, Real),
      cx: Vec(numTriangles, Real),
      cy: Vec(numTriangles, Real),
      weight: Real,
    },
  ],
  Real,
  ({ ax, ay, bx, by, cx, cy, weight }) => {
    const equilateral = sum(
      vec(numTriangles, Real, (i) => {
        const a = [ax[i], ay[i]];
        const b = [bx[i], by[i]];
        const c = [cx[i], cy[i]];

        const ab = norm(vsub(b, a));
        const bc = norm(vsub(c, b));
        const ca = norm(vsub(a, c));

        return add(
          sqr(sub(area(ab, bc, ca), (size / 10) ** 2)),
          add(add(sqr(sub(ab, bc)), sqr(sub(bc, ca))), sqr(sub(ca, ab))),
        );
      }),
    );
    const canvas = sum(
      vec(numTriangles, Real, (i) => {
        return [
          [ax[i], ay[i]],
          [bx[i], by[i]],
          [cx[i], cy[i]],
        ]
          .flatMap((v) => v.map((x) => sqr(min(0, min(x, sub(size, x))))))
          .reduce(add);
      }),
    );
    const disjoint = sum(
      vec(numTriangles, Real, (i) => {
        const a1 = [ax[i], ay[i]];
        const b1 = [bx[i], by[i]];
        const c1 = [cx[i], cy[i]];
        const p = select(
          clockwise(a1, b1, c1),
          Vec(3, Vec(2, Real)),
          [c1, b1, a1],
          [a1, b1, c1],
        );
        return sum(
          vec(numTriangles, Real, (j) => {
            const a2 = [neg(ax[j]), neg(ay[j])];
            const b2 = [neg(bx[j]), neg(by[j])];
            const c2 = [neg(cx[j]), neg(cy[j])];
            const q = select(
              clockwise(a1, b1, c1),
              Vec(3, Vec(2, Real)),
              [c2, b2, a2],
              [a2, b2, c2],
            );
            return select(
              ileq(numTriangles, i, j),
              Real,
              0,
              sqr(max(0, neg(sdPolygon(minkowskiSum(p, q), [0, 0])))),
            );
          }),
        );
      }),
    );
    return add(equilateral, mul(weight, add(canvas, disjoint)));
  },
);

const compiled = await compile(
  fn(
    [
      {
        ax: Vec(numTriangles, Real),
        ay: Vec(numTriangles, Real),
        bx: Vec(numTriangles, Real),
        by: Vec(numTriangles, Real),
        cx: Vec(numTriangles, Real),
        cy: Vec(numTriangles, Real),
        weight: Real,
      },
    ],
    {
      z: Real,
      ax: Vec(numTriangles, Real),
      ay: Vec(numTriangles, Real),
      bx: Vec(numTriangles, Real),
      by: Vec(numTriangles, Real),
      cx: Vec(numTriangles, Real),
      cy: Vec(numTriangles, Real),
    },
    (inputs) => {
      const { ret, grad } = vjp(lagrangian)(inputs);
      const { ax, ay, bx, by, cx, cy } = grad(1);
      return { z: ret, ax, ay, bx, by, cx, cy };
    },
  ),
);

const grad: optimizer.Fn = (
  x: Float64Array,
  weight: number,
  dx: Float64Array,
): number => {
  const { z, ax, ay, bx, by, cx, cy } = compiled({
    ax: x.subarray(0, numTriangles) as any,
    ay: x.subarray(numTriangles, numTriangles * 2) as any,
    bx: x.subarray(numTriangles * 2, numTriangles * 3) as any,
    by: x.subarray(numTriangles * 3, numTriangles * 4) as any,
    cx: x.subarray(numTriangles * 4, numTriangles * 5) as any,
    cy: x.subarray(numTriangles * 5, numTriangles * 6) as any,
    weight,
  });

  // https://github.com/rose-lang/rose/issues/111
  dx.set(ax as any, 0);
  dx.set(ay as any, numTriangles);
  dx.set(bx as any, numTriangles * 2);
  dx.set(by as any, numTriangles * 3);
  dx.set(cx as any, numTriangles * 4);
  dx.set(cy as any, numTriangles * 5);

  return z;
};

const init = (seed: string): Triangle[] => {
  const rng = seedrandom(seed);
  const triangles: Triangle[] = [];
  for (let i = 0; i < numTriangles; ++i) {
    const a = [rng() * size, rng() * size];
    const b = [rng() * size, rng() * size];
    const c = [rng() * size, rng() * size];
    triangles.push({ a, b, c });
  }
  return triangles;
};

const serialize = (triangles: Triangle[]): Float64Array => {
  const x = new Float64Array(numTriangles * 6);
  for (let i = 0; i < numTriangles; ++i) {
    const { a, b, c } = triangles[i];
    x[i] = a[0];
    x[i + numTriangles] = a[1];
    x[i + numTriangles * 2] = b[0];
    x[i + numTriangles * 3] = b[1];
    x[i + numTriangles * 4] = c[0];
    x[i + numTriangles * 5] = c[1];
  }
  return x;
};

const deserialize = (x: Float64Array): Triangle[] => {
  const triangles: Triangle[] = [];
  for (let i = 0; i < numTriangles; ++i) {
    const a = [x[i], x[i + numTriangles]];
    const b = [x[i + numTriangles * 2], x[i + numTriangles * 3]];
    const c = [x[i + numTriangles * 4], x[i + numTriangles * 5]];
    triangles.push({ a, b, c });
  }
  return triangles;
};

const optimize = (triangles: Triangle[]): Triangle[] => {
  const x = serialize(triangles);
  let params = optimizer.start(numTriangles * 6);
  while (params.optStatus !== "EPConverged")
    params = optimizer.stepUntil(grad, x, params, () => false);
  return deserialize(x);
};

const hueFactor = 360 / numTriangles;

const svg = (triangles: Triangle[]): string => {
  const lines = [
    `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${size} ${size}">`,
  ];
  for (let i = 0; i < numTriangles; ++i) {
    const { a, b, c } = triangles[i];
    const hue = i * hueFactor;
    lines.push(
      `  <polygon points="${a} ${b} ${c}" fill="hsl(${hue} 50% 50%)" />`,
    );
  }
  lines.push("</svg>", "");
  return lines.join("\n");
};

const main = () => {
  const original = init("");
  const optimized = optimize(original);
  Bun.write("out/triangles.svg", svg(optimized));
};

main();

File: triangles.jl
module Lbfgs
# adapted from https://github.com/penrose/penrose/blob/0ab136a32e8e8d7df2dc896d5702b499a6b4b594/packages/core/src/engine/Lbfgs.ts

import LinearAlgebra.dot

@kwdef struct Config
  m
  armijo
  wolfe
  min_interval
  max_steps
  epsd
end

@kwdef struct SY
  s
  y
end

@kwdef struct State
  x
  grad
  s_y
end

function line_search!(cfg, f, x0, r, fx0, grad, x)
  duf_at_x0 = -dot(r, grad)

  a = 0
  b = Inf
  t = 1
  j = 0

  while true
    x .= x0 - t * r

    if abs(a - b) < cfg.min_interval || j > cfg.max_steps
      break
    end

    fx = f(x, grad)
    is_armijo = fx <= fx0 + cfg.armijo * t * duf_at_x0
    is_wolfe = -dot(r, grad) >= cfg.wolfe * duf_at_x0

    if !is_armijo
      b = t
    elseif !is_wolfe
      a = t
    else
      break
    end

    if b < Inf
      t = (a + b) / 2
    else
      t = 2 * a
    end

    j += 1
  end

  t
end

function first_step!(cfg, f, x)
  T = eltype(x)
  n = length(x)
  x0 = copy(x)

  grad = zeros(T, n)
  fx = f(x, grad)

  r = copy(grad)
  line_search!(cfg, f, x0, r, fx, grad, x)

  State(x=x0, grad=r, s_y=[])
end

@kwdef struct Info
  state
  fx
  r
  x
  t
end

function step_until!(cfg, f, x, state, stop)
  T = eltype(x)
  n = length(x)
  grad = zeros(T, n)

  rho = zeros(T, cfg.m)
  alpha = zeros(T, cfg.m)
  q = zeros(T, n)
  r = zeros(T, n)

  while true
    fx = f(x, grad)

    if length(state.s_y) < cfg.m
      push!(state.s_y, SY(s=x - state.x, y=grad - state.grad))
    else
      (; s, y) = last(state.s_y)
      s .= x - state.x
      y .= grad - state.grad
    end
    pushfirst!(state.s_y, pop!(state.s_y))

    state.x .= x
    state.grad .= grad

    for (j, pair) in enumerate(state.s_y)
      s_j = pair.s
      y_j = pair.y
      rho[j] = 1 / (dot(y_j, s_j) + cfg.epsd)
    end

    q .= grad

    for (j, pair) in enumerate(state.s_y)
      s_j = pair.s
      y_j = pair.y
      alpha_j = rho[j] * dot(s_j, q)
      alpha[j] = alpha_j
      q -= alpha_j * y_j
    end

    pair = first(state.s_y)
    s_k = pair.s
    y_k = pair.y
    gamma = dot(s_k, y_k) / (dot(y_k, y_k) + cfg.epsd)
    r .= gamma * q

    for (j, pair) in Iterators.reverse(enumerate(state.s_y))
      s_j = pair.s
      y_j = pair.y
      alpha_j = alpha[j]
      beta = rho[j] * dot(y_j, r)
      r += s_j * (alpha_j - beta)
    end

    t = line_search!(cfg, f, state.x, r, fx, grad, x)

    if stop(Info(state=state, fx=fx, r=r, x=x, t=t))
      break
    end
  end
end

end

using LinearAlgebra: dot, norm
import Random
import Zygote

unconstrained_converged(norm_grad) = norm_grad < 1e-2

function ep_converged(x0, x1, fx0, fx1)
  ep_stop = 1e-3
  abs(fx1 - fx0) < ep_stop || norm(x1 - x0) < ep_stop
end

# adapted from https://github.com/penrose/penrose/blob/0ab136a32e8e8d7df2dc896d5702b499a6b4b594/packages/core/src/engine/Optimizer.ts
function exterior_point!(f, x)
  cfg = Lbfgs.Config(
    m=17,
    armijo=0.001,
    wolfe=0.9,
    min_interval=1e-9,
    max_steps=10,
    epsd=1e-11,
  )
  last_x = copy(x)
  last_fx = nothing
  weight = 1e3
  g = (x, dx) -> f(x, weight, dx)
  while true
    fx = nothing
    state = Lbfgs.first_step!(cfg, g, x)
    Lbfgs.step_until!(cfg, g, x, state, info -> begin
      fx = info.fx
      unconstrained_converged(dot(info.state.grad, info.r))
    end)
    # this forces at least one round, but the original optimizer forced two
    if !isnothing(last_fx) && ep_converged(last_x, x, last_fx, fx)
      break
    end
    last_x .= x
    last_fx = fx
    weight *= 10
  end
end

function reorder(p)
  n = length(p)
  i = argmin((v[2], v[1]) for v in p)
  j = ((i - 1) % n) + 1
  [p[i:n]; p[1:i-1]; p[i:i]; p[j:j]]
end

cross(u, v) = u[1] * v[2] - u[2] * v[1]

# https://cp-algorithms.com/geometry/minkowski.html
function minkowski_sum(left, right)
  m = length(left)
  n = length(right)
  p = reorder(left)
  q = reorder(right)
  r = Zygote.Buffer([])
  i = 1
  j = 1
  while i <= m || j <= n
    push!(r, p[i] + q[j])
    z = cross(p[i+1] - p[i], q[j+1] - q[j])
    if z >= 0 && i <= m
      i += 1
    end
    if z <= 0 && j <= n
      j += 1
    end
  end
  copy(r)
end

# https://iquilezles.org/articles/distfunctions2d/
function sd_polygon(v, p)
  N = length(v)
  d = dot(p - v[1], p - v[1])
  s = 1.0
  j = N
  for i = 1:N
    e = v[j] - v[i]
    w = p - v[i]
    b = w - e * clamp(dot(w, e) / dot(e, e), 0.0, 1.0)
    d = min(d, dot(b, b))
    c = [p[2] >= v[i][2], p[2] < v[j][2], e[1] * w[2] > e[2] * w[1]]
    if all(c) || all(.!c)
      s *= -1.0
    end
    j = i
  end
  s * sqrt(d)
end

@kwdef struct Triangle
  a
  b
  c
end

chunks(serialized) = Iterators.partition(serialized, 6)

deserialize(chunk) = Triangle(
  a=chunk[1:2],
  b=chunk[3:4],
  c=chunk[5:6],
)

area(ab, bc, ca) =
  0.25 * sqrt(
    (ab + bc + ca) * (-ab + bc + ca) * (ab - bc + ca) * (ab + bc - ca)
  )

# https://algs4.cs.princeton.edu/91primitives/
clockwise((; a, b, c)) =
  (b[1] - a[1]) * (c[2] - a[2]) < (c[1] - a[1]) * (b[2] - a[2])

function lagrangian(serialized, weight)
  obj = sum(chunks(serialized)) do chunk
    (; a, b, c) = deserialize(chunk)
    ab = norm(b - a)
    bc = norm(c - b)
    ca = norm(a - c)
    (area(ab, bc, ca) - 0.01)^2 + (ab - bc)^2 + (bc - ca)^2 + (ca - ab)^2
  end
  constrs = sum(chunks(serialized)) do left
    t1 = deserialize(left)
    p = clockwise(t1) ? [t1.c, t1.b, t1.a] : [t1.a, t1.b, t1.c]
    sum(chunks(serialized)) do right
      t2 = deserialize(right)
      q = clockwise(t2) ? [-t2.a, -t2.b, -t2.c] : [-t2.c, -t2.b, -t2.a]
      max(0, -sd_polygon(minkowski_sum(p, q), [0, 0]))^2
    end
  end
  obj + weight * constrs
end

function init(seed)
  Random.seed!(seed)
  triangles = []
  for _ = 1:10
    push!(triangles, Triangle(
      a=[rand(), rand()],
      b=[rand(), rand()],
      c=[rand(), rand()],
    ))
  end
  triangles
end

function val_and_grad(x, weight, dx)
  (; val, grad) = Zygote.withgradient(lagrangian, x, weight)
  dx .= first(grad)
  val
end

function optimize(triangles)
  cfg = Lbfgs.Config(
    m=17,
    armijo=0.001,
    wolfe=0.9,
    min_interval=1e-9,
    max_steps=10,
    epsd=1e-11,
  )
  x = vcat([vcat(t.a, t.b, t.c) for t in triangles]...)
  exterior_point!(val_and_grad, x)
  [deserialize(chunk) for chunk in chunks(x)]
end

vecstr(v) = join(v, ",")

function svg(triangles)
  lines = ["""<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1 1">"""]
  for (i, (; a, b, c)) in enumerate(triangles)
    points = "$(vecstr(a)) $(vecstr(b)) $(vecstr(c))"
    fill = "hsl($(i * 360 / length(triangles)) 50% 50%)"
    push!(lines, """  <polygon points="$points" fill="$fill" />""")
  end
  push!(lines, "</svg>")
  push!(lines, "")
  join(lines, "\n")
end

function main()
  original = init(0)
  optimized = optimize(original)
  write("out.svg", svg(optimized))
end

main()
